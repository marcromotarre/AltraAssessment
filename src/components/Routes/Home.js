import React, { useState } from 'react';
import { useSelector } from 'react-redux';
import styled from 'styled-components';

import {
  getUsersByName,
  getUsersByFilters,
  isAnyFilterApplied as checkAnyFilterApplied,
} from '../../utils/filters.js';
import MainBanner from '../Banners/MainBanner.js';
import UsersList from '../User/UsersList.js';
import SearchInput from '../SearchInput.js';
import FilterBox from '../Filter/FilterBox.js';
import DataManager from '../DataManager.js';

export const HomeContainer = styled.section``;

export const Header = styled.div`
  position: sticky;
  top: 0;
`;

const Home = () => {
  const { townName, users } = useSelector((state) => state.town);
  const { age, height, weight, hair_colors } = useSelector(
    (state) => state.filters
  );
  const [filteringMode, setFilteringMode] = useState(false);
  const [searchValue, setSearchValue] = useState('');

  const filteredUsers = getUsersByFilters({
    users,
    age,
    height,
    weight,
    hair_colors,
  });
  const searchedUsers = getUsersByName(filteredUsers, searchValue);

  const search = (inputSearchValue) => {
    setSearchValue(inputSearchValue);
  };

  const setFilteringModeTrue = () => {
    setFilteringMode(true);
  };

  const isAnyFilterApplied = checkAnyFilterApplied({
    age,
    height,
    weight,
    hair_colors,
  });

  return (
    <HomeContainer data-testid="HomeSite">
      <Header>
        <MainBanner
          townName={townName}
          onClickFilterIcon={setFilteringModeTrue}
          showFilterButton={!isAnyFilterApplied && !filteringMode}
        />
        <FilterBox
          filteredUsersNumber={filteredUsers.length}
          isAnyFilterApplied={isAnyFilterApplied}
          filteringMode={filteringMode}
          setFilteringMode={setFilteringMode}
        ></FilterBox>
        {!filteringMode && (
          <SearchInput onSearch={search} initValue={searchValue} />
        )}
      </Header>
      {!filteringMode && (
        <DataManager>
          <UsersList users={searchedUsers} usersPerPage={15} />
        </DataManager>
      )}
    </HomeContainer>
  );
};

export default Home;
