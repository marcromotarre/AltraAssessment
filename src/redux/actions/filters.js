import * as types from './../constants';

export const setAgeFilter = ({ filterData }) => ({
  type: types.SET_AGE_FILTER,
  payload: {
    filterData,
  },
});

export const setAgeMin = ({ min }) => ({
  type: types.SET_AGE_MIN,
  payload: {
    min,
  },
});

export const setAgeMax = ({ max }) => ({
  type: types.SET_AGE_MAX,
  payload: {
    max,
  },
});

export const setAgeRange = ({ range }) => ({
  type: types.SET_AGE_RANGE,
  payload: {
    range,
  },
});

export const setHeightFilter = ({ filterData }) => ({
  type: types.SET_HEIGHT_FILTER,
  payload: {
    filterData,
  },
});

export const setHeightMin = ({ min }) => ({
  type: types.SET_HEIGHT_MIN,
  payload: {
    min,
  },
});

export const setHeightMax = ({ max }) => ({
  type: types.SET_HEIGHT_MAX,
  payload: {
    max,
  },
});

export const setHeightRange = ({ range }) => ({
  type: types.SET_HEIGHT_RANGE,
  payload: {
    range,
  },
});

export const setWeightFilter = ({ filterData }) => ({
  type: types.SET_WEIGHT_FILTER,
  payload: {
    filterData,
  },
});

export const setWeightMin = ({ min }) => ({
  type: types.SET_WEIGHT_MIN,
  payload: {
    min,
  },
});

export const setWeightMax = ({ max }) => ({
  type: types.SET_WEIGHT_MAX,
  payload: {
    max,
  },
});

export const setWeightRange = ({ range }) => ({
  type: types.SET_WEIGHT_RANGE,
  payload: {
    range,
  },
});

export const addHairColorFilter = ({ color }) => ({
  type: types.ADD_HAIR_COLOR_FILTER,
  payload: {
    color,
  },
});

export const deleteHairColorFilter = ({ color }) => ({
  type: types.DELETE_HAIR_COLOR_FILTER,
  payload: {
    color,
  },
});

export const clearHairColorsFilters = () => ({
  type: types.CLEAR_HAIR_COLORS_FILTER,
});
