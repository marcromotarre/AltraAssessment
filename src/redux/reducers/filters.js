import * as types from './../constants';

export const initialState = {
  age: {
    min: '',
    max: '',
    range: true,
  },
  weight: {
    min: '',
    max: '',
    range: true,
  },
  height: {
    min: '',
    max: '',
    range: true,
  },
  hair_colors: [],
};

export function reducer(state = initialState, { type, payload }) {
  switch (type) {
    case types.SET_AGE_FILTER:
      return {
        ...state,
        age: payload.filterData,
      };
    case types.SET_AGE_MIN:
      return {
        ...state,
        age: { ...state.age, min: payload.min },
      };
    case types.SET_AGE_MAX:
      return {
        ...state,
        age: { ...state.age, max: payload.max },
      };
    case types.SET_AGE_RANGE:
      return {
        ...state,
        age: { ...state.age, range: payload.range },
      };
    case types.SET_HEIGHT_FILTER:
      return {
        ...state,
        height: payload.filterData,
      };
    case types.SET_HEIGHT_MIN:
      return {
        ...state,
        height: { ...state.height, min: payload.min },
      };
    case types.SET_HEIGHT_MAX:
      return {
        ...state,
        height: { ...state.height, max: payload.max },
      };
    case types.SET_HEIGHT_RANGE:
      return {
        ...state,
        height: { ...state.height, range: payload.range },
      };
    case types.SET_WEIGHT_FILTER:
      return {
        ...state,
        weight: payload.filterData,
      };
    case types.SET_WEIGHT_MIN:
      return {
        ...state,
        weight: { ...state.weight, min: payload.min },
      };
    case types.SET_WEIGHT_MAX:
      return {
        ...state,
        weight: { ...state.weight, max: payload.max },
      };
    case types.SET_WEIGHT_RANGE:
      return {
        ...state,
        weight: { ...state.weight, range: payload.range },
      };
    case types.ADD_HAIR_COLOR_FILTER:
      return {
        ...state,
        hair_colors: !state.hair_colors.includes(payload.color)
          ? [...state.hair_colors, payload.color]
          : [...state.hair_colors],
      };

    case types.DELETE_HAIR_COLOR_FILTER:
      return {
        ...state,
        hair_colors: state.hair_colors.filter(
          (color) => color !== payload.color
        ),
      };

    case types.CLEAR_HAIR_COLORS_FILTER:
      return {
        ...state,
        hair_colors: [],
      };

    default:
      return {
        ...state,
      };
  }
}
