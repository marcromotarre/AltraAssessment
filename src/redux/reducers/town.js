import * as types from './../constants';
import {
  getTownsNames,
  getUsersByTown,
  getUniqueHairColorsByTown,
} from '../../utils/town';

export const initialState = {
  pending: false,
  townName: '',
  users: [],
  uniqueHairColors: [],
  error: null,
};

export function reducer(state = initialState, { type, payload }) {
  switch (type) {
    case types.FETCH_TOWN_PENDING:
      return {
        ...state,
        pending: true,
      };
    case types.FETCH_TOWN_SUCCESS:
      const townName = getTownsNames(payload.response)[0];
      return {
        ...state,
        pending: false,
        townName,
        users: getUsersByTown(payload.response, townName),
        uniqueHairColors: getUniqueHairColorsByTown(payload.response, townName),
      };
    case types.FETCH_TOWN_ERROR:
      return {
        ...state,
        pending: false,
        error: payload.error,
      };

    default:
      return {
        ...state,
      };
  }
}
