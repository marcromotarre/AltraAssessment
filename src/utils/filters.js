export const isMinMaxFilterApplied = (filter) =>
  filter.min === '' && filter.max === '' ? false : true;

export const isArrayFilterApplied = (filter) =>
  filter.length > 0 ? true : false;

export const getFilterMinMaxString = (data) => {
  if (data.range && data.min !== '' && data.max !== '') {
    return `${data.min} - ${data.max}`;
  } else if (data.min !== '') {
    return `${data.min}`;
  } else if (data.max !== '') {
    return `${data.max}`;
  } else {
    return '';
  }
};

export const isAgeOrWeightOrHeightFilterApplied = ({ age, height, weight }) => {
  return isMinMaxFilterApplied(age) ||
    isMinMaxFilterApplied(height) ||
    isMinMaxFilterApplied(weight)
    ? true
    : false;
};

export const isAnyFilterApplied = ({ age, height, weight, hair_colors }) => {
  return isAgeOrWeightOrHeightFilterApplied({ age, height, weight }) ||
    isArrayFilterApplied(hair_colors)
    ? true
    : false;
};

export const getUsersByName = (users, name) => {
  return users.filter((user) =>
    user.name.toLowerCase().includes(name.toLowerCase())
  );
};

export const getUsersByProfession = (users, profession) => {
  return users.filter((user) => user.professions.includes(profession));
};

export const getUniqueHairColors = (users) => {
  return users
    .map((user) => user.hair_color)
    .filter((value, index, self) => self.indexOf(value) === index);
};

export const getUsersByMinMaxFilter = (users, type, filter) => {
  if (!isMinMaxFilterApplied(filter)) {
    return [...users];
  } else {
    if (filter.min !== '' && filter.max !== '' && filter.range) {
      return users.filter((user) => {
        return (
          user[type] >= parseFloat(filter.min) &&
          user[type] <= parseFloat(filter.max)
        );
      });
    } else if (
      (filter.min !== '' && filter.max !== '' && !filter.range) ||
      filter.min !== ''
    ) {
      return users.filter((user) => {
        return Math.round(user[type]) === parseFloat(filter.min);
      });
    } else {
      return users.filter((user) => {
        return Math.round(user[type]) === parseFloat(filter.max);
      });
    }
  }
};

export const getUsersByArrayFilter = (users, type, filter) => {
  if (!isArrayFilterApplied(filter)) {
    return [...users];
  } else {
    return users.filter((user) => {
      return filter.includes(user[type]);
    });
  }
};

export const getUsersByFilters = ({
  users = [],
  age = { min: '', max: '', range: false },
  height = { min: '', max: '', range: false },
  weight = { min: '', max: '', range: false },
  hair_colors = [],
}) => {
  return getUsersByArrayFilter(
    getUsersByMinMaxFilter(
      getUsersByMinMaxFilter(
        getUsersByMinMaxFilter(users, 'age', age),
        'weight',
        weight
      ),
      'height',
      height
    ),
    'hair_color',
    hair_colors
  );
};
