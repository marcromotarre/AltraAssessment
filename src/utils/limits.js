export const getMinUserValueByField = (users, field) => {
  return users ? Math.min(...users.map((user) => Math.floor(user[field]))) : 0;
};

export const getMaxUserValueByField = (users, field) => {
  return users ? Math.max(...users.map((user) => Math.ceil(user[field]))) : 0;
};
