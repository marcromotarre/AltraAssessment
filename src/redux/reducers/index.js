import { combineReducers } from 'redux';
import { reducer as town } from './town';
import { reducer as filters } from './filters';

export default combineReducers({
  town,
  filters,
});
