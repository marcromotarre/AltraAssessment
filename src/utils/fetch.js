import fetch from 'cross-fetch';
import * as actions from './../redux/actions/town';

export function fetchTown(dispatch) {
  dispatch(actions.fetchTownPending());
  fetch(
    'https://raw.githubusercontent.com/rrafols/mobile_test/master/data.json'
  )
    .then((response) => response.json())
    .then((response) => {
      if (response.error) {
        throw response.error;
      }
      dispatch(actions.fetchTownSuccess(response));
    })
    .catch((error) => {
      dispatch(actions.fetchTownError(error));
    });
}
