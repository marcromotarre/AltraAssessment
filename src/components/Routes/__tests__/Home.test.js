import React from 'react';
import { renderRouteRedux } from '../../../utils/test';
import users from '../../../fixtures/users';
import '@testing-library/jest-dom/extend-expect';
import { initialState as filtersInitialState } from './../../../redux/reducers/filters';

import { Router } from '@reach/router';

import Home from '../Home';

describe('Home Route Component', () => {
  it('Home Route Snapshot', () => {
    expect(
      renderRouteRedux(
        <Router>
          <Home path="/" />
        </Router>,
        {
          route: '/',
          initialState: {
            filters: filtersInitialState,
            town: { users: users },
          },
        }
      ).baseElement
    ).toMatchSnapshot();
  });
});
