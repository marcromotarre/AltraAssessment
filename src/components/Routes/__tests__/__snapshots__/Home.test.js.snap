// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Route Component Home Route Snapshot 1`] = `
<body>
  <div>
    <div
      style="outline: none;"
      tabindex="-1"
    >
      <section
        class="sc-fzoxKX ja-DdBK"
        data-testid="HomeSite"
      >
        <div
          class="sc-fzoKki dDxpXH"
        >
          <div
            class="sc-AxjAm dCtgNG"
          >
            <div
              class="sc-AxirZ bnKdnV"
            >
              <svg
                class="sc-AxiKw iKXCa"
              />
            </div>
            
            <button
              class="sc-AxhCb bjuFrh"
            >
              <svg
                width="21.5"
              />
            </button>
          </div>
          <section
            class="sc-fznWOq bjYIJR"
          >
            <div
              class="sc-fzolEj ePmwvw"
            />
          </section>
          <div
            class="sc-fznyAO bTufgS true"
          >
            <svg
              class="sc-fznKkj dylAVn"
              data-testid="search-icon"
              width="21.42"
            />
            <input
              class="sc-fznZeY lblHde"
              data-testid="search-input"
              value=""
            />
          </div>
        </div>
        <div
          class="sc-fzplWN iRwTAx"
        >
          <a
            href="/user/0"
          >
            <section
              class="sc-AxheI cuTdVv"
            >
              <img
                class="sc-Axmtr hcomXe"
                data-testid="user-image"
                src="image1"
              />
              <div
                class="sc-AxmLO cIzxDa"
              >
                <label
                  class="sc-fzozJi gpkgCw"
                  data-testid="user-name"
                >
                  one
                </label>
                <div
                  class="sc-fzoLsD livjYH"
                  data-testid="user-properties"
                >
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      width="14.63"
                    />
                     
                    <label>
                      10
                    </label>
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      height="17.7"
                    />
                    <label>
                      50
                    </label>
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      width="17.7"
                    />
                    <label>
                      40
                    </label>
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      class="sc-AxgMl OQqCX"
                      color="Red"
                      width="15.26"
                    />
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      width="14.34"
                    />
                    <label>
                      3
                    </label>
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      width="15.26"
                    />
                     
                    <label>
                      2
                    </label>
                  </div>
                </div>
              </div>
            </section>
          </a>
          <a
            href="/user/1"
          >
            <section
              class="sc-AxheI cuTdVv"
            >
              <img
                class="sc-Axmtr hcomXe"
                data-testid="user-image"
                src="image2"
              />
              <div
                class="sc-AxmLO cIzxDa"
              >
                <label
                  class="sc-fzozJi gpkgCw"
                  data-testid="user-name"
                >
                  two
                </label>
                <div
                  class="sc-fzoLsD livjYH"
                  data-testid="user-properties"
                >
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      width="14.63"
                    />
                     
                    <label>
                      15
                    </label>
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      height="17.7"
                    />
                    <label>
                      70
                    </label>
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      width="17.7"
                    />
                    <label>
                      40
                    </label>
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      class="sc-AxgMl grTrYX"
                      color="Green"
                      width="15.26"
                    />
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      width="14.34"
                    />
                    <label>
                      1
                    </label>
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      width="15.26"
                    />
                     
                    <label>
                      1
                    </label>
                  </div>
                </div>
              </div>
            </section>
          </a>
          <a
            href="/user/2"
          >
            <section
              class="sc-AxheI cuTdVv"
            >
              <img
                class="sc-Axmtr hcomXe"
                data-testid="user-image"
                src="image3"
              />
              <div
                class="sc-AxmLO cIzxDa"
              >
                <label
                  class="sc-fzozJi gpkgCw"
                  data-testid="user-name"
                >
                  three
                </label>
                <div
                  class="sc-fzoLsD livjYH"
                  data-testid="user-properties"
                >
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      width="14.63"
                    />
                     
                    <label>
                      40
                    </label>
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      height="17.7"
                    />
                    <label>
                      120
                    </label>
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      width="17.7"
                    />
                    <label>
                      100
                    </label>
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      class="sc-AxgMl OQqCX"
                      color="Red"
                      width="15.26"
                    />
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      width="14.34"
                    />
                    <label>
                      2
                    </label>
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      width="15.26"
                    />
                     
                    <label>
                      2
                    </label>
                  </div>
                </div>
              </div>
            </section>
          </a>
          <a
            href="/user/3"
          >
            <section
              class="sc-AxheI cuTdVv"
            >
              <img
                class="sc-Axmtr hcomXe"
                data-testid="user-image"
                src="image4"
              />
              <div
                class="sc-AxmLO cIzxDa"
              >
                <label
                  class="sc-fzozJi gpkgCw"
                  data-testid="user-name"
                >
                  four
                </label>
                <div
                  class="sc-fzoLsD livjYH"
                  data-testid="user-properties"
                >
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      width="14.63"
                    />
                     
                    <label>
                      17
                    </label>
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      height="17.7"
                    />
                    <label>
                      30
                    </label>
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      width="17.7"
                    />
                    <label>
                      20
                    </label>
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      class="sc-AxgMl kVJfuU"
                      color="Pink"
                      width="15.26"
                    />
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      width="14.34"
                    />
                    <label>
                      0
                    </label>
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      width="15.26"
                    />
                     
                    <label>
                      1
                    </label>
                  </div>
                </div>
              </div>
            </section>
          </a>
          <a
            href="/user/4"
          >
            <section
              class="sc-AxheI cuTdVv"
            >
              <img
                class="sc-Axmtr hcomXe"
                data-testid="user-image"
                src="image4"
              />
              <div
                class="sc-AxmLO cIzxDa"
              >
                <label
                  class="sc-fzozJi gpkgCw"
                  data-testid="user-name"
                >
                  Five
                </label>
                <div
                  class="sc-fzoLsD livjYH"
                  data-testid="user-properties"
                >
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      width="14.63"
                    />
                     
                    <label>
                      80
                    </label>
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      height="17.7"
                    />
                    <label>
                      30
                    </label>
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      width="17.7"
                    />
                    <label>
                      35
                    </label>
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      class="sc-AxgMl OQqCX"
                      color="Red"
                      width="15.26"
                    />
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      width="14.34"
                    />
                    <label>
                      1
                    </label>
                  </div>
                  <div
                    class="sc-AxhUy gPWWVZ"
                    data-testid="user-property"
                  >
                    <svg
                      width="15.26"
                    />
                     
                    <label>
                      0
                    </label>
                  </div>
                </div>
              </div>
            </section>
          </a>
        </div>
      </section>
    </div>
  </div>
</body>
`;
